src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,467 @@
+import { ResourceStore } from './ResourceStore';
+import streamifyArray from 'streamify-array';
+import { UnsupportedMediaTypeHttpError } from '../util/errors/UnsupportedMediaTypeHttpError';
+import { CONTENT_TYPE_QUADS, DATA_TYPE_BINARY, DATA_TYPE_QUAD } from '../util/ContentTypes';
+import { createReadStream, createWriteStream, promises as fsPromises, Stats } from 'fs';
+import { ensureTrailingSlash, trimTrailingSlashes } from '../util/Util';
+import { extname, join as joinPath, normalize as normalizePath } from 'path';
+
+/**
+ * Resource store storing its data in the file system backend.
+ * All requests will throw an {@link NotFoundHttpError} if unknown identifiers get passed.
+ */
+export class FileResourceStore implements ResourceStore {
+  private readonly baseRequestURI: string;
+  private readonly rootFilepath: string;
+  private readonly interactionController: InteractionController;
+  private readonly metadataController: MetadataController;
+
+  /**
+   * @param baseRequestURI - Will be stripped of all incoming URIs and added to all outgoing ones to find the relative
+   * path.
+   * @param rootFilepath - Root filepath in which the resources and containers will be saved as files and directories.
+   * @param interactionController - Instance of InteractionController to use.
+   * @param metadataController - Instance of MetadataController to use.
+   */
+  public constructor(baseRequestURI: string, rootFilepath: string, interactionController: InteractionController,
+    metadataController: MetadataController) {
+    this.baseRequestURI = trimTrailingSlashes(baseRequestURI);
+    this.rootFilepath = trimTrailingSlashes(rootFilepath);
+    this.interactionController = interactionController;
+    this.metadataController = metadataController;
+  }
+
+  /**
+   * Store the incoming data as a file under a file path corresponding to `container.path`,
+   * where slashes correspond to subdirectories.
+   * @param container - The identifier to store the new data under.
+   * @param representation - Data to store. Only File streams are supported.
+   *
+   * @returns The newly generated identifier.
+   */
+  public async addResource(container: ResourceIdentifier, representation: Representation): Promise<ResourceIdentifier> {
+    if (representation.dataType !== DATA_TYPE_BINARY) {
+      throw new UnsupportedMediaTypeHttpError('FileResourceStore only supports binary representations.');
+    }
+
+    // Get the path from the request URI, all metadata triples if any, and the Slug and Link header values.
+    const path = this.parseIdentifier(container);
+    const { slug, raw } = representation.metadata;
+    const linkTypes = representation.metadata.linkRel?.type;
+    let metadata;
+    if (raw.length > 0) {
+      metadata = this.metadataController.generateReadableFromQuads(raw);
+    }
+
+    const isContainer = this.interactionController.isContainer(slug, linkTypes);
```

Comment:
Add one comment line above explaining what the "block" does.

(My code style would be to have no newlines within the block then, but that's up to you. Here's for instance a random example of the "code blocks with leading comment" style: https://github.com/LDflex/LDflex/blob/89225e1cf3b87814bd8addd4ea609128c204f810/src/PathFactory.js#L10-L39 I picked that one, because you most likely have no idea what this code is for, but you can probably still follow it.)